### PR TITLE
Workspace for Qbft Message Definition

### PR DESCRIPTION
**Pull request by rain-on**
Signed-off-by: Trent Mohay <trent.mohay@consensys.net>

<!-- Thanks for sending a p

### CODE REVIEW

**Review by saltiniroberto — COMMENTED** [review]
Added a few comments.

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/messages/Messages.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+  public class PreparedCertificate {
+    // proposal payload
+    public ConsensusRoundIdentifier roundIdentifier;
+    public Hash digest;
+    public Signature signature; // this is the signature of the block and view
+    public List<Prepare> receivedPrepareMsgs;
+  }
+
+  public static class RoundChangePayload {
```
[the remark this answers is not in the document]
Comment (rain-on) [reply consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/messages/Messages.java:47]: My bad -this was meant to represent IBFT2 (which I think this does)

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/messages/Messages.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+  ////////////////////////////////////////////////////////////////
+
+  ////////////////////////////////////////////////////////////////
+  // MESSAGES
+
+  // Each message is RLP encoded (in their declared order)
+
+  public static class Proposal {
+    public ConsensusRoundIdentifier view;
+    public Hash blockHash;
```
[the remark this answers is not in the document]
Comment (rain-on) [reply consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/messages/Messages.java:65]: again - meant to represent IBFT2

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/messages/Messages.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    public Block block;
+    public Optional<RoundChangeCertificate> blockJustification;
```
[the remark this answers is not in the document]
Comment (rain-on) [reply consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/messages/Messages.java:67]: again - ibft2

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/messages/Messages.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
```
[the remark this answers is not in the document]
Comment (rain-on) [reply consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/messages/Messages.java:68]: again ibft2

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/messages/Messages.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    public Signature commitSeal; // the signature of the block being sealed
+
+    // this is the signature of the keccak256 hash of the rlp encoded bytes of preceding items
+    public Signature signature;
+  }
+
+  public static class RoundChange {
+    public RoundChangePayload payload;
+    public Optional<Block> proposedBlock;
+
```
[the remark this answers is not in the document]
Comment (rain-on) [reply consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/messages/Messages.java:93]: ibf2

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/messages/Messages.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    // this is the signature of the keccak256 hash of the rlp encoded bytes of preceding items
+    public Signature signature;
```
[the remark this answers is not in the document]
Comment (rain-on) [reply consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/messages/Messages.java:95]: ibft2

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/payload/PreparedRoundMetadata.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.hyperledger.besu.consensus.qbft.payload;
+
+import org.hyperledger.besu.ethereum.core.Hash;
+import org.hyperledger.besu.ethereum.rlp.RLPInput;
+import org.hyperledger.besu.ethereum.rlp.RLPOutput;
+
+public class PreparedRoundMetadata {
```
Comment (saltiniroberto) [inline consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/payload/PreparedRoundMetadata.java:21]: I don't see equals and hashCode implemented here.
Does it matter?

**Review by saltiniroberto — COMMENTED** [review]
As far as the message structure is concerned, it LGTM.

PreparedRoundMetadata may be missing equals and hashCode though.

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/payload/PreparedRoundMetadata.java (hunk 7 above)
Comment (rain-on) [reply consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/payload/PreparedRoundMetadata.java:21]: yeah probably need it - but tbh - we really need to know that these types represent the correct message structures

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/messagewrappers/Proposal.java [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    rlpOut.endList();
+    return rlpOut.encoded();
+  }
+
+  public static Proposal decode(final Bytes data) {
+    final RLPInput rlpIn = RLP.input(data);
+    rlpIn.enterList();
+    final SignedData<ProposalPayload> payload =
+        PayloadSerializers.readSignedProposalPayloadFrom(rlpIn);
+    final List<SignedData<PreparePayload>> prepares =
```
Comment (jframe) [inline consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/messagewrappers/Proposal.java:68]: According to the spec the prepares should be the last field included, not the 2nd field.

Comment (rain-on) [reply consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/messagewrappers/Proposal.java]: done

**Comment by rain-on** [comment]
Superseded by other QBFT PRs.

**Pull request closed without merging**